### PR TITLE
[Notifier] Add AllMySms Bridge

### PR DESCRIPTION
**Pull request by qdequippe**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/14739
| Recipe | https://github.com/symfony/recipes/pull/865

Add AllMySms bridge to Symfony Notifier

### CODE REVIEW

**Comment by qdequippe** [comment]
@OskarStark bridge added ;)

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        $this->login = $login;
+        $this->apiKey = $apiKey;
+        $this->from = $from;
+
+        parent::__construct($client, $dispatcher);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('allmysms://%s?from=%s', $this->getEndpoint(), $this->from);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php]: From parameter should only be added if not null

File: src/Symfony/Component/Notifier/Bridge/AllMySms/LICENSE [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2019-2020 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/LICENSE]:
```suggestion
Copyright (c) 2020 Fabien Potencier
```

**Review by OskarStark — APPROVED** [review]
Wow perfect 🤩 just 2 minor comments

**Comment by OskarStark** [comment]
Could you please add it to the documentation + add a recipe? 😎

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransportFactory.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        $scheme = $dsn->getScheme();
+
+        if ('allmysms' !== $scheme) {
+            throw new UnsupportedSchemeException($dsn, 'allmysms', $this->getSupportedSchemes());
+        }
+
+        $login = $this->getUser($dsn);
+        $apiKey = $this->getPassword($dsn);
+        $from = $dsn->getOption('from');
+
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransportFactory.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.2.0
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/CHANGELOG.md]:
```suggestion
5.3.0
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php (hunk 1 above)
Comment (qdequippe) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php]: I have to remove `?from=%s`?

**Comment by qdequippe** [comment]
> Could you please add it to the documentation + add a recipe?

Yes of course! Where to add the documentation? in https://github.com/symfony/symfony-docs?

**Comment by OskarStark** [comment]
Yes you can search for other bridges, there is an open PR for GatewayAPI bridge, your should be the same, same for recipe in Symfony/recipes

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        if (201 !== $response->getStatusCode()) {
+            $error = $response->toArray(false);
+
+            throw new TransportException(sprintf('Unable to send the SMS: "%s" (%s).', $error['description'], $error['code']), $response);
+        }
+
+        $success = $response->toArray(false);
+
+        $sentMessage = new SentMessage($message, (string) $this);
+        $sentMessage->setMessageId($success['smsId']);
```
Comment (ismail1432) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php:86]: maybe check that `$success['smsId']` is defined like it's done here https://github.com/symfony/google-chat-notifier/blob/5.x/GoogleChatTransport.php#L133 
```php
if(null === $smsId = $success['smsId'] ?? null) {
throw error

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SentMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Quentin Dequippe <quentin@dequippe.tech>
+ *
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * @experimental in 5.3
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransportFactory.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+
+/**
+ * @author Quentin Dequippe <quentin@dequippe.tech>
+ *
+ * @experimental in 5.3
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransportFactory.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        $this->login = $login;
+        $this->apiKey = $apiKey;
+        $this->from = $from;
+
+        parent::__construct($client, $dispatcher);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('allmysms://%s', $this->getEndpoint());
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/AllMySmsTransport.php:49]: If from is set we should output it in the toString method

File: src/Symfony/Component/Notifier/Bridge/AllMySms/README.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+-----------
+
+```
+ALLMYSMS_DSN=allmysms://LOGIN:APIKEY@default?from=FROM
+```
+
+where:
+- `LOGIN` is your user ID
+- `APIKEY` is your AllMySms API key
+- `FROM` is your sender (optional). Default: 36180
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/README.md]:
```suggestion
- `FROM` is your sender (optional, default: 36180)
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        return new AllMySmsTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield [
+            'allmysms://host.test',
+            'allmysms://login:apiKey@host.test',
+        ];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php:33]: Please add one test here where you provide from parameter

File: src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+     * @return AllMySmsTransport
+     */
+    public function createTransport(?HttpClientInterface $client = null): TransportInterface
+    {
+        return new AllMySmsTransport('accountSid', 'authToken', 'from', $client ?: $this->createMock(HttpClientInterface::class));
+    }
+
+    public function toStringProvider(): iterable
+    {
+        yield ['allmysms://api.allmysms.com', $this->createTransport()];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php:34]: Please add a test where you provide from parameter

File: src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+use Symfony\Component\Notifier\Tests\TransportTestCase;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+final class AllMySmsTransportTest extends TransportTestCase
+{
+    /**
+     * @return AllMySmsTransport
+     */
+    public function createTransport(?HttpClientInterface $client = null): TransportInterface
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php]:
```suggestion
    public function createTransport(?HttpClientInterface $client = null, string $from = null): TransportInterface
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        return new AllMySmsTransport('accountSid', 'authToken', 'from', $client ?: $this->createMock(HttpClientInterface::class));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportTest.php]:
```suggestion
        return new AllMySmsTransport('accountSid', 'authToken', $from, $client ?: $this->createMock(HttpClientInterface::class));
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        return new AllMySmsTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield [
+            'allmysms://host.test',
+            'allmysms://login:apiKey@host.test',
+            'allmysms://login:apiKey@host.test?from=TEST',
+        ];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php:38]: I think you need to add another yield

Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php]:
```suggestion
        yield [
            'allmysms://host.test',
            'allmysms://login:apiKey@host.test',
        ];
        
        yield [
            'allmysms://host.test?from=TEST',
            'allmysms://login:apiKey@host.test?from=TEST',
        ];        
```

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php]: This should fix the test

Comment (qdequippe) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/Tests/AllMySmsTransportFactoryTest.php]: Oops sorry you're right thx!

File: src/Symfony/Component/Notifier/Bridge/AllMySms/LICENSE [hunk 16]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2020 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/LICENSE]:
```suggestion
Copyright (c) 2021 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/README.md [hunk 17]
```diff
@@ -0,0 +1,24 @@
+AllMySms Notifier
+===============
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/README.md]:
```suggestion
AllMySms Notifier
=================
```

File: src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        },
+        {
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
+        "symfony/notifier": "^5.3"
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json:21]:
```suggestion
        "symfony/notifier": "~5.3.0"
```

**Review by OskarStark — APPROVED** [review]
Good to go afterwards 👍

File: src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json (hunk 18 above)
Comment (derrabus) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json:21]: `^5.3` is correct.

Comment (OskarStark) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json:21]: Ok 👍  I am a bit confused, but yes you are right 👍

**Review by OskarStark — APPROVED** [review]
Thank you 👍

File: src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json (hunk 18 above)
Comment (derrabus) [reply src/Symfony/Component/Notifier/Bridge/AllMySms/composer.json:21]: Yes, this was kind-of a back and forth here. Sorry for the confusion.

**Review by OskarStark — APPROVED** [review]

**Comment by OskarStark** [comment]
Thank you Quentin.

**Pull request merged**